src/sources/docker.rs

```diff
@@ -719,46 +756,160 @@ impl ContainerLogInfo {
                         return None;
                     }
                 }
-                // Supply timestamp
-                log_event.insert(event::TIMESTAMP.clone(), timestamp.with_timezone(&Utc));
 
                 self.last_log = Some((timestamp, self.generation));
 
                 let log = splitter.next()?;
                 let remove_len = message.len() - log.len();
                 bytes_message.advance(remove_len);
+
+                // Provide the timestamp.
+                Some(timestamp.with_timezone(&Utc))
             }
             Err(error) => {
                 // Recieved bad timestamp, if any at all.
                 error!(message="Didn't recieve rfc3339 timestamp from docker",%error);
-                // So log whole message
+                // So continue normally but without a timestamp.
+                None
             }
-        }
+        };
 
-        // Message is actually one line from stderr or stdout, and they are delimited with newline
-        // so that newline needs to be removed.
-        if bytes_message
+        // Message is actually one line from stderr or stdout, and they are
+        // delimited with newline, so that newline needs to be removed.
+        // If there's no newline, the event is considered partial, and will
+        // either be merged within the docker source, or marked accordingly
+        // before sending out, depending on the configuration.
+        let is_partial = if bytes_message
             .last()
             .map(|&b| b as char == '\n')
             .unwrap_or(false)
         {
             bytes_message.truncate(bytes_message.len() - 1);
-        }
+            false
+        } else {
+            true
+        };
 
-        // Supply message
+        // Populate the message. We only need a message to merge.
+        let mut log_event = Event::new_empty_log().into_log();
         log_event.insert(event::MESSAGE.clone(), bytes_message.freeze());
 
-        // Supply container
-        log_event.insert(CONTAINER.clone(), self.id.0.clone());
+        // We got the message, that's the only part of the event we merge, so if
+        // we have the automatic partial event merging enabled, we might be able
+        // to return early and skip the logic of parsing the rest of the fields.
+        // Otherwize we continue populating the event with the rest of the
+        // fields and then we might want to consume it to initialize a partial
+        // merge event state - or return it.
+        let use_event_to_initialize_partial_merge_event_state = if auto_partial_merge {
```

Comment:
I found the logic in this whole section somewhat complicated to follow, so I made a truth table. There are essentially four boolean values we're working with, two actual `bool`s and two `Option`s (the merge `state` and marker `field`) which are considered true if they're `Some`.

Based on those values, we determine which actions to take. The three things we do in various combinations are `merge` the new event into our state, `initialize` the new event with all its fields, `mark` the event as partial via the configured field, and `return` the event or state.

| auto_partial_merge | is_partial | state | field | actions
|------|---------|-------|-------|-------                   
| Y    | Y       | Y     | -     | merge
| Y    | Y       | N     | -     | initialize, merge
| Y    | N       | Y     | -     | merge, return
| Y    | N       | N     | -     | initialize, return
| N    | -       | Y     | -     | !! INVALID !!
| N    | Y       | N     | Y     | initialize, mark, return
| N    | Y       | N     | N     | initialize, return
| N    | N       | N     | -     | initialize, return

Notes:
* There can never be state if auto merging is off
* The marker field only matters if auto merging is off

First, the fact that our model (the two configurable values and state) can represent an invalid state is a little bit of a smell and suggests that maybe we could use an enum of some kind to more accurately reflect the possible cases.

Second, and this may just be a matter of preference so feel free to disagree, but the current code is a bit difficult to hold in your head due to the more procedural style. As you read, it's important to remember which sets of conditions you've already handled over the past ~100 LoC. To me, code that is written more explicitly to match against each input pattern is easier to grok and more obviously exhaustive. Often this can just be a big `match` statement that's a direct translation of the truth table.

Reply:
I completely agree that this is way too complicated. The truth table doesn't capture one of the critical parts of this design - we skip filling the event with fields if we don't have to, thus saving some CPU cycles. This is the root cause of the complexity, and also a result of the optimization to save those CPU cycles on filling the fields that will be unused.
I've actually had a version that's much more readable and maintainable just before I implemented this optimization - https://github.com/timberio/vector/pull/1504/commits/52b604219c99f96ff6c0fe2dffae4a28f11b17e6.

This feedback clearly confirms my initial suspicion that this extra step will increase the maintenance cost too much - so what I suggest is to do a step back, accept that we won't be caring about those extra fields being filled and just stick to the algorithm version from https://github.com/timberio/vector/pull/1504/commits/52b604219c99f96ff6c0fe2dffae4a28f11b17e6. It's much easier to comprehend and change if needed.

I think it's still good that we tried this optimized version of the algorithm, and saw what tradeoffs it can offer. Later on, if we revisit this, we'll know the cost of this.

Alternatively, we can probably improve the algorithm further, as suggested, by using enums and more advanced state management descriptors, but I think even that would render just minor improvements, and the complexity would be still unreasonably high.

Reply:
@lukesteensen if like going back to the the logic at https://github.com/timberio/vector/commit/52b604219c99f96ff6c0fe2dffae4a28f11b17e6 - I can change adapt the code today, so we can move forward with this PR :smile:

Reply:
> The truth table doesn't capture one of the critical parts of this design - we skip filling the event with fields if we don't have to, thus saving some CPU cycles.

This is what I meant by `initialize` in the actions column.

> This is the root cause of the complexity, and also a result of the optimization to save those CPU cycles on filling the fields that will be unused. I've actually had a version that's much more readable and maintainable just before I implemented this optimization

That sounds good! The optimization would be nice, but I agree we've learned the complexity doesn't seem to be worthwhile right now. If we have any issues with the performance we can always revisit after profiling confirms it would be a win.